state-sync/src/coordinator.rs

```diff
@@ -1146,69 +1227,64 @@ impl<T: ExecutorProxyTrait> StateSyncCoordinator<T> {
-    /// * issue a new request if too much time passed since requesting highest_synced_version + 1.
-    fn check_progress(&mut self) {
-        if self.request_manager.no_available_peers() {
-            return;
-        }
-
+    /// * Kick starts the initial sync process (e.g., syncing to a waypoint or target).
+    /// * Issues a new request if too much time has passed since the last request was sent.
+    fn check_progress(&mut self) -> Result<(), Error> {
         if self.is_consensus_executing() {
-            return; // No need to send out chunk requests: consensus is executing.
+            return Err(Error::ConsensusIsExecuting);
         }
 
-        // check that we made progress in fulfilling consensus sync request
-        let sync_request_expired = self.sync_request.as_ref().map_or(false, |req| {
-            let default_timeout = Duration::from_millis(self.config.sync_request_timeout_ms);
-            if let Some(tst) = req.last_progress_tst.checked_add(default_timeout) {
-                return SystemTime::now().duration_since(tst).is_ok();
-            }
-            false
-        });
-        // notify consensus if sync request timed out
-        if sync_request_expired {
-            counters::SYNC_REQUEST_RESULT
-                .with_label_values(&[counters::TIMEOUT_LABEL])
-                .inc();
-            warn!(LogSchema::event_log(
-                LogEntry::SyncRequest,
-                LogEvent::Timeout
-            ));
+        // Check if the sync request has timed out (i.e., if we aren't committing fast enough)
+        if let Some(sync_request) = self.sync_request.as_ref() {
+            let timeout_between_commits =
+                Duration::from_millis(self.config.sync_request_timeout_ms);
+            let commit_deadline = sync_request
+                .last_commit_timestamp
+                .checked_add(timeout_between_commits)
+                .ok_or_else(|| {
+                    Error::IntegerOverflow("The commit deadline timestamp has overflown!".into())
+                })?;
+            if SystemTime::now().duration_since(commit_deadline).is_ok() {
+                // Timeout exceeded!
+                counters::SYNC_REQUEST_RESULT
+                    .with_label_values(&[counters::TIMEOUT_LABEL])
+                    .inc();
+                warn!(LogSchema::event_log(
+                    LogEntry::SyncRequest,
+                    LogEvent::Timeout
+                ));
 
-            if let Some(sync_request) = self.sync_request.take() {
-                if let Err(e) = Self::send_sync_req_callback(
-                    sync_request,
-                    Err(format_err!("request timed out")),
-                ) {
-                    error!(
-                        LogSchema::event_log(LogEntry::SyncRequest, LogEvent::CallbackFail)
-                            .error(&e)
-                    );
+                if let Some(sync_request) = self.sync_request.take() {
```

Comment:
Put a comment here saying what this blob does, like starts a new request.

I'm not quite sure what all the refactoring in this function is doing, there's a lot of nesting in here now

Reply:
This sends a callback to consensus. The structure of this code is more a function of how Rust handles as_ref() and take() for options. But, I'll add a comment to help reduce the confusion.